api/operator/v1alpha1/istiocsr_types.go

```diff
@@ -0,0 +1,280 @@
+// istio-csr agent and keeps it in the desired state.
+//
+// +kubebuilder:object:root=true
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:printcolumn:name="GRPC Endpoint",type="string",JSONPath=".status.istioCSRGRPCEndpoint"
+// +kubebuilder:validation:XValidation:rule="self.metadata.name == 'default'",message="istiocsr is a singleton, .metadata.name must be 'default'"
+// +operator-sdk:csv:customresourcedefinitions:displayName="IstioCSR"
+type IstioCSR struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the IstioCSR.
+	// +kubebuilder:validation:Required
+	// +required
+	Spec IstioCSRSpec `json:"spec,omitempty"`
+
+	// status is the most recently observed status of the IstioCSR.
+	Status IstioCSRStatus `json:"status,omitempty"`
+}
+
+// IstioCSRSpec is the specification of the desired behavior of the IstioCSR.
+type IstioCSRSpec struct {
+	// istioCSRConfig is for configuring the istio-csr agent behavior.
+	// +kubebuilder:validation:Required
+	// +required
+	IstioCSRConfig *IstioCSRConfig `json:"istioCSRConfig,omitempty"`
+
+	// controllerConfig is for configuring the controller for setting up
+	// defaults to enable istio-csr agent.
+	ControllerConfig *ControllerConfig `json:"controllerConfig,omitempty"`
+}
+
+// IstioCSRConfig is for configuring the istio-csr agent behavior.
+type IstioCSRConfig struct {
+	// logLevel supports value range as per [kubernetes logging guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use).
+	// +kubebuilder:default:=1
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=5
+	// +kubebuilder:validation:Optional
+	// +optional
+	LogLevel int32 `json:"logLevel,omitempty"`
+
+	// logFormat is for specifying the output format of istio-csr agent logging.
+	// Support log formats are text and json.
+	// +kubebuilder:default:=text
+	// +kubebuilder:validation:Optional
+	// +optional
+	LogFormat string `json:"logFormat,omitempty"`
+
+	// certManager is for configuring cert-manager specifics.
+	// +kubebuilder:validation:Required
+	// +required
+	CertManager *CertManagerConfig `json:"certManager,omitempty"`
+
+	// istiodTLSConfig is for configuring istiod certificate specifics.
+	// +kubebuilder:validation:Required
+	// +required
+	IstiodTLSConfig *IstiodTLSConfig `json:"istiodTLSConfig,omitempty"`
+
+	// server is for configuring the server endpoint used by istio
+	// for obtaining the certificates.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Server *ServerConfig `json:"server,omitempty"`
+
+	// istio is for configuring the istio specifics.
+	// +kubebuilder:validation:Required
+	// +required
```

Comment:
thanks for adding both in each of:
```
// +kubebuilder:validation:Required
// +required
```
experimented a bit through it (and reading over https://github.com/zecke/Kubernetes/blob/master/docs/devel/api-conventions.md#optional-vs-required), concluded that for now as long as we have to use `kubebuilder:validation:Required|Optional` markers now because we're using `sigs.k8s.io/controller-tools v0.15.0`.

Once we bump to v0.16+ (we cannot bump to 0.16 right away as it is based on k8s 1.31 so it can only be done during next rebase) use of only `+required`/`+optional` would suffice.